### ISSUE TITLE
sd-resolve: getaddrinfo hangs

### ISSUE THREAD

**Issue by LionNatsu** [systemd/systemd#9162]
**systemd version the issue has been seen with**
> 238

**Used distribution**
 > AOSC OS

**Expected behaviour you didn't see**
 > systemd-timesyncd stops normally.

**Unexpected behaviour you saw**
 > systemd-timesyncd hangs ten seconds or even longer to stop when the NTP cannot be resolved correctly, e.g. DNS cannot be reached.

**Steps to reproduce the problem**
timesyncd.conf:

    [Time]
    NTP=axrfav.,avfra

Steps to reproduce:

    systemctl restart systemd-timesyncd  # apply the new config
    systemctl stop    systemd-timesyncd  # hang

**Comment by LionNatsu** [comment]
Run `/usr/lib/systemd/systemd-timesyncd` and Ctrl+C can also trigger the bug.

Using `ltrace`, I looked into the code. After sending the interrupt signal, the main thread sent termination packet to the worker thread.
```
(timesyncd-manager.c) manager_free => (sd-resolve.c) sd_resolve_unref => resolve_free => pthread_join
```
The main thread was blocked and waiting for the worker thread.

While the work thread is also blocking by...
```
(sd-resolve.c) thread_worker => handle_request => getaddrinfo
```
That means sd-resolve cannot exit until the last getaddrinfo request returned.

In this case, timesyncd always blocks the shutdown progress. It's so hard to wait.

**Comment by poettering** [comment]
if getaddrinfo() hangs we can't clean up its thread. There's little we can do about that really I fear...

**Comment by LionNatsu** [comment]
Indeed. I think we may isolate the call to a separated thread and `pthread_cancel()` it when the caller wants to clean up. Working threads communicate to theirs `getaddrinfo()` threads with arguments and `join()`.

How about asynchronous version? `getaddrinfo_a()` can be cancelled easily, but asynchronous calls is in very different way comparing with current  implementation of sd-resolve.

**Comment by LionNatsu** [comment]
The main point is, `pthread_cancel()` terminates a thread forcibly. Though terminating the working threads is not elegant or safe(maybe), we can terminate only the `getaddrinfo()` call if we separate it from the working thread.

**Comment by poettering** [comment]
pthread_cancel() is not a usable API IRL. We invoke foreign code through the NSS module, which is likely to allocate resources while it runs, such as memory or file descriptors. If we'd cancel the thread in the middle, then the thread might go away, but these resources would remain allocated.

What we probably could do to make this better is possibly to rework sd-resolve so that the context object is reference counted, and each thread keeps a ref. Then, when the user drops a ref we could permit that instantly, but the threads would possibly partially stay around until their NSS calls complete. And when they do we'd clean up the context too. This means we'd not leak resources, and we could instantly release context objects from the user side, but possibly threads would stay around in the background for a while until they finish off.

**Comment by LionNatsu** [comment]
That's pretty nice, but it's still a problem if threads linger for too long.

FYI, I found the combination of `getaddrinfo_a(GAI_NOWAIT, …)`(do queries) and `gai_suspend(…, timeout)`(wait for results) can set a timeout for queries. This may help threads run out from the blocking state and check if it should exit or wait again.
